server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java

```diff
@@ -438,7 +451,29 @@ IndexFieldData.Builder fielddataBuilder(DenseVectorFieldType denseVectorFieldTyp
                     denseVectorFieldType.indexVersionCreated,
                     this,
                     denseVectorFieldType.dims,
-                    denseVectorFieldType.indexed
+                    denseVectorFieldType.indexed,
+                    denseVectorFieldType.indexVersionCreated.onOrAfter(NORMALIZE_COSINE)
+                        && denseVectorFieldType.indexed
+                        && denseVectorFieldType.similarity.equals(VectorSimilarity.COSINE) ? r -> new FilterLeafReader(r) {
+                            @Override
+                            public CacheHelper getCoreCacheHelper() {
+                                return r.getCoreCacheHelper();
+                            }
+
+                            @Override
+                            public CacheHelper getReaderCacheHelper() {
+                                return r.getReaderCacheHelper();
+                            }
+
+                            @Override
+                            public FloatVectorValues getFloatVectorValues(String fieldName) throws IOException {
+                                FloatVectorValues values = in.getFloatVectorValues(fieldName);
```

Comment:
Do we need to check if `values` is null? E.g. if this segment doesn't have vectors?